### ISSUE TITLE
[REQUEST]: Support building no_std binaries

### ISSUE THREAD

**Issue by piegamesde** [compiler-explorer/compiler-explorer#3274]
### Is your feature request related to a problem? Please describe

I've managed to compile code for different `no_std` targets successfully, but as soon as I check "Compile to binary" then I get linker errors.

[Example 1](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:20,positionColumn:1,positionLineNumber:20,selectionStartColumn:1,selectionStartLineNumber:20,startColumn:1,startLineNumber:20),source:'%23!!%5Ballow(dead_code)%5D%0A%23!!%5Bno_std%5D%0A%23!!%5Bno_main%5D%0A%23!!%5Bfeature(lang_items)%5D%0A%0Apub+fn+main()+%7B%0A++++unsafe+%7B%0A++++++++core::ptr::copy_nonoverlapping(0xe0000000+as+*const+u8,+0x80000000+as+*mut+u8,+256+*+1024)%3B%0A++++%7D%0A++++loop+%7B%7D%0A%7D%0A%0A%23%5Blang+%3D+%22eh_personality%22%5D%0Aextern+%22C%22+fn+eh_personality()+%7B%7D%0A%0A%0A%23%5Bpanic_handler%5D%0Afn+panic(_panic:+%26core::panic::PanicInfo%3C!'_%3E)+-%3E+!!+%7B%0A++++loop+%7B%7D%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:51.07528619046543,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:nightly,filters:(b:'0',binary:'0',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D3',selection:(endColumn:45,endLineNumber:3,positionColumn:45,positionLineNumber:3,selectionStartColumn:45,selectionStartLineNumber:3,startColumn:45,startLineNumber:3),source:1,tree:'1'),l:'5',n:'0',o:'rustc+nightly+(Rust,+Editor+%231,+Compiler+%231)',t:'0'),(h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',tree:'1',wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+nightly+(Compiler+%231)',t:'0')),k:48.92471380953459,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

[Example 2](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:24,positionColumn:1,positionLineNumber:24,selectionStartColumn:1,selectionStartLineNumber:24,startColumn:1,startLineNumber:24),source:'%23!!%5Bno_main%5D%0A%23!!%5Bno_std%5D%0A%23!!%5Ballow(dead_code)%5D%0A%0Apub+fn+foo()+%7B%0A++++bar()%3B%0A%7D%0A%0A%23%5Binline(always)%5D%0Apub+fn+bar()+%7B%0A++++unsafe+%7B%0A++++++++core::ptr::copy_nonoverlapping(0xe0000000+as+*const+u8,+0x80000000+as+*mut+u8,+256+*+1024)%3B%0A++++%7D%0A++++loop+%7B%7D%0A%7D%0A%0Ause+core::panic::PanicInfo%3B%0A%23%5Bpanic_handler%5D%0A%23%5Bno_mangle%5D%0Apub+fn+panic(_info:+%26PanicInfo)+-%3E+!!+%7B%0A%09unreachable!!()%3B%0A%7D%0A%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:51.07528619046543,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:nightly,filters:(b:'0',binary:'0',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'--target%3Driscv32imc-unknown-none-elf+-C+opt-level%3Ds',selection:(endColumn:45,endLineNumber:3,positionColumn:45,positionLineNumber:3,selectionStartColumn:45,selectionStartLineNumber:3,startColumn:45,startLineNumber:3),source:1,tree:'1'),l:'5',n:'0',o:'rustc+nightly+(Rust,+Editor+%231,+Compiler+%231)',t:'0'),(h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',tree:'1',wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+nightly+(Compiler+%231)',t:'0')),k:48.92471380953459,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

### Describe the solution you'd like

My main motivation is that I want to debug the code generation of LLVM's `memcpy` generation. When not building as a standalone binary, the compiler usually only generates a jump/call to the respective intrinsic function so I cannot have a look at it.

### Describe alternatives you've considered

I also tried building binaries for `std` targets. This should work, however the amount of generated boilerplate code is too way too large so it's not really a viable option.

### Additional context

Maybe this is already supported and I'm too dumb to get it to work.

**Comment by dkm** [comment]
Logging output here:

```
error: linking with `/opt/compiler-explorer/gcc-11.1.0/bin/gcc` failed: exit status: 1
  |
  = note: "/opt/compiler-explorer/gcc-11.1.0/bin/gcc" "-m64" "/app/output.example.c8e30779-cgu.0.rcgu.o" "-Wl,--as-needed" "-L" "/opt/compiler-explorer/rust-nightly/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/opt/compiler-explorer/rust-nightly/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-2a6a2797f7a73818.rlib" "/opt/compiler-explorer/rust-nightly/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0e3656b1fda5fd7b.rlib" "/opt/compiler-explorer/rust-nightly/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-f98cce041f593917.rlib" "-Wl,-Bdynamic" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/opt/compiler-explorer/rust-nightly/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/app/output.s" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs"
  = note: /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /lib/x86_64-linux-gnu/Scrt1.o: in function `_start':
          (.text+0x16): undefined reference to `__libc_csu_fini'
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: (.text+0x1d): undefined reference to `__libc_csu_init'
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: (.text+0x24): undefined reference to `main'
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: (.text+0x2a): undefined reference to `__libc_start_main'
          collect2: error: ld returned 1 exit status
          
  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: aborting due to previous error

Compiler returned: 1
```


```
error: linking with `/opt/compiler-explorer/gcc-11.1.0/bin/gcc` failed: exit status: 1
  |
  = note: "/opt/compiler-explorer/gcc-11.1.0/bin/gcc" "/app/output.example.c8e30779-cgu.0.rcgu.o" "-Wl,--as-needed" "-L" "/opt/compiler-explorer/rust-nightly/lib/rustlib/riscv32imc-unknown-none-elf/lib" "-Wl,-Bstatic" "/opt/compiler-explorer/rust-nightly/lib/rustlib/riscv32imc-unknown-none-elf/lib/librustc_std_workspace_core-abe67a39589d1521.rlib" "/opt/compiler-explorer/rust-nightly/lib/rustlib/riscv32imc-unknown-none-elf/lib/libcore-181d7fed250fef4a.rlib" "/opt/compiler-explorer/rust-nightly/lib/rustlib/riscv32imc-unknown-none-elf/lib/libcompiler_builtins-49ef1a39314020d6.rlib" "-Wl,-Bdynamic" "-Wl,-znoexecstack" "-L" "/opt/compiler-explorer/rust-nightly/lib/rustlib/riscv32imc-unknown-none-elf/lib" "-o" "/app/output.s" "-Wl,--gc-sections" "-no-pie" "-nodefaultlibs"
  = note: /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /app/output.example.c8e30779-cgu.0.rcgu.o: relocations in generic ELF (EM: 243)
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /app/output.example.c8e30779-cgu.0.rcgu.o: relocations in generic ELF (EM: 243)
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /app/output.example.c8e30779-cgu.0.rcgu.o: relocations in generic ELF (EM: 243)
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /app/output.example.c8e30779-cgu.0.rcgu.o: relocations in generic ELF (EM: 243)
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /app/output.example.c8e30779-cgu.0.rcgu.o: relocations in generic ELF (EM: 243)
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /app/output.example.c8e30779-cgu.0.rcgu.o: relocations in generic ELF (EM: 243)
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /app/output.example.c8e30779-cgu.0.rcgu.o: relocations in generic ELF (EM: 243)
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /app/output.example.c8e30779-cgu.0.rcgu.o: relocations in generic ELF (EM: 243)
          /opt/compiler-explorer/gcc-11.1.0/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: /app/output.example.c8e30779-cgu.0.rcgu.o: error adding symbols: file in wrong format
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error

Compiler returned: 1
```

Looks like you are missing some lib in the command line. Adding `-lc` remove some errors, but not all. I suspect there's something needed to ask the linker to use PIC versions of libs. I'll have a look later if nobody gives the solution

**Comment by dkm** [comment]
I guess cargo may be using extra args when calling rustc?

Here's the first one: https://godbolt.org/z/4MovsoGPP
The second one seems to be gcc picking the wrong lib, leading to incorrect relocation in resulting program.

**Comment by piegamesde** [comment]
Hm, it compiles now, but it generates a call to `memcpy@GLIBC_2.14`. I'm not sure because I've never done x86_64 `no_std` before, but I'd expect libc to not be available in that context, forcing the usage of compiler intrinsics instead.

**Comment by dkm** [comment]
That's really up to `rustc`, I'm not sure compiler-explorer can do anything about that... Maybe try to see how `rustc` is called when creating a no_std app on linux using cargo, maybe there's some extra option ?

**Comment by piegamesde** [comment]
Yes, you're right. The system architecture is probably `x86_64-unknown-linux-gnu`, and we'd need something like `x86_64-unknown-none` instead. This however means that my approach in Example 1 cannot work and I need to get some embedded target to work to properly test the code generation. (Also, this means that my issue actually is two mostly independent issues)

I tried using `aarch64-unknown-none` as target in Example 2, the error remains the same: `error adding symbols: file in wrong format`.

**Comment by dkm** [comment]
Sorry, looked at it a bit too quickly. Your 2nd example shows that rustc is calling host's linker instead of cross linker, hence the error message. I guess something's missing for `rustc` to call the correct cross tools.

**Comment by dkm** [comment]
....which by the way will be a little difficult as we don't really support this kind of things... So basically, cross-linking is not supported currently, you would need to also pick a version for gcc/ld...

**Comment by piegamesde** [comment]
Yeah. I normally link with `lld`, where I only need to make sure that it supports the desired target. For linking with `ld`, I think one needs to have the matching cross-gcc installed and set the binary name prefix via a variable.

**Comment by dkm** [comment]
How do you ask `rustc` to use `lld` or `GNU ld` ? Is it looking in `$PATH` for what is available and then pick something ? Or do you force it using something else ?

**Comment by piegamesde** [comment]
I'm unsure. I think the default is set by the compilation target, but something like `-C linker=clang` and/or `-C link-arg=-fuse-ld=lld` should override it. But I wouldn't know where to add these flags. The Cargo configuration calls these `rustflags`, so they're added to `rustc` calls I guess?
